Add `database_game` support for recording inference results (`record_predictions`, default off).

17zhangw opened the pull request:
Add `database_game` support for recording inference results (`record_predictions`, default off).

This PR extends support for recording inference results (alongside true execution costs) in the Flask microservice by passing `record_predictions` (default off) as an argument to `database_game`. The following information is captured:

1. The current workload query being explored;
2. The current action state when costing happened;
3. The predicted cost from the microservice;
4. The prediction data from the microservice (this includes debug information);
5. The true cost from running the query against the database;
6. Whether a true cost was obtained successfully or not (e.g., if we hit a unique key violation and fail).

Currently, cost is defined as elapsed time in microseconds. 
The comparison is between the models' predictions of how long a query will take VS how long a query actually took.

- To record: pass `--record_predictions` to `database_game`, either directly or via `doit action_recommendation --database_game_args="--use_microservice --record_predictions"`.
- To view the recorded results: `localhost:5000` will show the inference results by default. See the screenshot on GitHub.
    - For more advanced viewing, see `behavior/microservice/app.py` -- you can issue raw SQL queries to the underlying SQLite `./artifacts/behavior/microservice/inference.db` backing it all, or you can open that directly in SQLite.

![image](https://user-images.githubusercontent.com/8349071/153101298-673ce8e4-bc0b-4233-9e80-00b6e68e581a.png)

---

Extends support for recording inference results (in comparison to true execution costs) with the Flask microservice. Currently, we write down the following pieces of information. These pieces are only written down if `record_predictions` is passed as an argument to the database game.
1. the current workload query being explored
2. the current action state when costing happened
3. the predicted cost from the microservice
4. the prediction data from the microservice (this includes debug information)
5. the true cost from running the query against the database
6. whether a true cost was obtained successfully or not (i.e., if we hit a unique key violation and fail)

Currently, `localhost:5000/prediction_results` will return a whole dump of all the objects.
![Screenshot from 2022-02-02 16-22-38](https://user-images.githubusercontent.com/7199435/152260307-fab65f3d-c827-4496-89f0-a2a5752492d1.png)

Two things that are still WIP / food for thought:
~~1. All this data might benefit from being stored into a postgres pilot table instance (along with other state we want to scrape)~~
~~2. Put the data into a table to make it easier to look at~~

17zhangw committed:
Infrastructure to note down predictions / true statistics in the pilot

17zhangw committed:
Stashing query predictions in sqlite3 db

17zhangw committed:
ci_python

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -220,9 +222,22 @@ struct ExplainMicroserviceCost {
+        {
+          // If we're recording results, then append this particular inference result to the accumulator.
+          // We'll be able to identify this failure since the model_cost_ will show a low prediction.
+          inference_results_ << (res->body.c_str()) << std::endl;
```
Since the `if` and `else` code paths are identical, move it to after the if (with your documentation)?

lmwnshn replied:
Since the `if` and `else` code paths are identical, move it to after the if (with your documentation)?

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -97,7 +99,7 @@ struct ExplainCost {
 struct ExplainMicroserviceCost {
  public:
-  explicit ExplainMicroserviceCost(httplib::Client *client, pqxx::result *rset) {
+  explicit ExplainMicroserviceCost(httplib::Client *client, pqxx::result *rset, bool record_results) {
```
Consistency

```suggestion
  explicit ExplainMicroserviceCost(httplib::Client *client, pqxx::result *rset, bool record_predictions) {
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -114,7 +116,7 @@ struct ExplainMicroserviceCost {
         for (auto &plan_node : doc.GetArray()) {
           rapidjson::Document::Object &&plan_obj = plan_node["Plan"].GetObject();
-          Cost(client, plan_obj);
+          Cost(client, plan_obj, record_results);
```
```suggestion
          Cost(client, plan_obj, record_predictions);
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -156,7 +158,7 @@ struct ExplainMicroserviceCost {
   }
 
-  void Cost(httplib::Client *client, rapidjson::Document::Object &node) {
+  void Cost(httplib::Client *client, rapidjson::Document::Object &node, bool record_results) {
```
```suggestion
  void Cost(httplib::Client *client, rapidjson::Document::Object &node, bool record_predictions) {
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -233,14 +248,15 @@ struct ExplainMicroserviceCost {
       for (auto &child_node : node["Plans"].GetArray()) {
         rapidjson::Document::Object &&child = child_node.GetObject();
-        Cost(client, child);
+        Cost(client, child, record_results);
```
```suggestion
        Cost(client, child, record_predictions);
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -233,14 +248,15 @@ struct ExplainMicroserviceCost {
  public:
   // The model cost is aggregated by making network requests.
   double model_cost_ = 0;
+  std::stringstream inference_results_;
```
Please add a comment describing what this is, it isn't clear what a `stringstream` of "inference results" would mean.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
 bool DatabaseState::IsTerminal() const { return finished_; }
 
+template<class T>
+void DatabaseState::ApplyHistory(T &txn, bool use_hypopg) const {
```
Why is this templated? Is it because there is no base that would work for both txn and subtxn? If so, please add as a static assert what expected template types are. Would it work with a `pqxx::dbtransaction`?

lmwnshn replied:
I am also slightly uncomfortable with:

1. the potential explosion of `use_blah` parameters,
2. the way that this signature can give callers the wrong impression (`DatabaseState` already contains `game_`).

I think you are passing in `use_hypopg` because you want to override the `game_`'s setting in some cases, right? What if instead we do

```
void DatabaseState<T>::ApplyHistory(T &txn);
void DatabaseState<T>::CustomApplyHistory(T &txn, bool use_hypopg, ...); // do as you wish
```

to make the distinction clear? and the former would simply invoke the latter, pulling out parameters from `game_` automatically.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
+}
+
+std::string DatabaseState::GetAppliedStateRepresentation() const {
+  // This function is used to return a string describing the current state of the database
```
What is different from the many existing ToString methods that are already part of OpenSpiel? Is a custom function outside of the framework necessary / should this be an override for one of those instead?

```
  std::string ToString() const override;
  std::string InformationStateString(Player player) const override;
  std::string ObservationString(Player player) const override;
```

lmwnshn replied:
Currently, it just seems like duplicate functionality.

lmwnshn commented on action/selection/open_spiel/database.h:
```diff
@@ -66,6 +66,16 @@ class DatabaseState : public State {
+  // Gets the actual runtime cost with EXPLAIN(ANALYZE, BUFFERS) of executing query
+  // under the given transaction. The function returns a pair of <success, time>.
+  template<class T>
+  std::pair<bool, double> GetExplainAnalyzeCostUs(T &txn, const std::string &query) const;
```
Nit: replace `std::pair<bool,double>` with a named struct. In general, I do not think we should use `pair` -- people get lazy and shove more `pair<pair<>>`.. stuff and you end up with `.first->second.first->second` nonsense.

lmwnshn replied:
Additionally, after looking at the implementation, you actually want a `std::optional<double>` not a `std::pair<bool, double>`. More accurately, you would probably want a `std::optional<Struct>`.

lmwnshn commented on action/selection/open_spiel/database.h:
```diff
@@ -66,6 +66,16 @@ class DatabaseState : public State {
+  template<class T>
+  void ApplyHistory(T &txn, bool use_hypopg) const;
+
+  std::string GetAppliedStateRepresentation() const;
```
See above comment in .cc; I do not think this needs to exist.

lmwnshn commented on action/selection/open_spiel/database_game.cc:
```diff
@@ -19,6 +19,7 @@ ABSL_FLAG(std::string, actions_path, "./actions.csv", "Path to CSV of possible S
 ABSL_FLAG(int, max_tuning_actions, 1, "Maximum number of tuning actions before the game ends.");
 ABSL_FLAG(bool, use_hypopg, true, "True if hypopg should be used for faking index builds.");
 ABSL_FLAG(bool, use_microservice, false, "True if microservice should be used for inference.");
+ABSL_FLAG(bool, record_predictions, false, "True if query inference results should be stored with the microservice");
```
```suggestion
ABSL_FLAG(bool, record_predictions, false, "True if query inference results should be stored in the microservice.");
```

lmwnshn commented on demo/mvp.sh:
```diff
@@ -72,6 +72,6 @@ PGPASSWORD=np_as_spiel_pass ./artifacts/noisepage/psql -h localhost -d np_as_spi
 
 # Start picking indexes.
-doit action_recommendation --database_game_args="--use_microservice"
+doit action_recommendation --database_game_args="--use_microservice --record_predictions"
```
```suggestion
doit action_recommendation --database_game_args="--use_microservice"
```

We probably do not want to record predictions by default, since as you mention this takes 1h on dev4.
(To be clear, what takes 1h? One iteration or all iterations of recommending indexes?)

lmwnshn replied:
To document for the future -- this was 1 iteration.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
+    // due to failing an integrity constraint check. ExplainMicroservice does not
+    // risk aborting since it doesn't actually execute a query.
+    //
+    // TODO(wz2): Need a graceful way of identifying and possibly modeling aborts
```
```suggestion
    // TODO(wz2): Need a graceful way of identifying and possibly modeling aborts.
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
+    ExplainAnalyzeCost cost{&rset};
+    return std::make_pair(true, cost.actual_total_time_ms_ * 1000);
+  } catch (const pqxx::integrity_constraint_violation& e) {
+    // Since we're re-executing a query, it is possible that the query might abort
```
The phrasing here is slightly weird. First line says we're re-executing (who did the original execution? are we always re-executing a query? This isn't true -- I might for example roll back the DB after collecting a workload trace.), then the last line says we're not actually executing a query. Additionally, this code might eventually get used by other callers, not necessarily `ExplainMicroservice`.

Consider something like,

```
EXPLAIN ANALYZE will execute the query that it is provided as part of the workload forecast.
However, this query may abort due to failing integrity constraint check(s) such as uniqueness.
We currently do not model nor handle aborts, therefore all aborted queries will be treated as
if they had zero cost.
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
+    // risk aborting since it doesn't actually execute a query.
+    //
+    // TODO(wz2): Need a graceful way of identifying and possibly modeling aborts
+    std::cerr << absl::StrCat("Encountered error executing query ", explain, " ", e.what()) << std::endl;
```
```suggestion
    std::cerr << absl::StrCat("ERROR while executing query ", explain, " ", e.what()) << std::endl;
```

either ERROR or WARN as you see fit

I suggest that we standardize on a set of prefixes (currently, only ERROR is used) to make it easy to grep and to debug from the index_picker script.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -483,28 +552,19 @@ std::vector<double> DatabaseState::Returns() const {
-        pqxx::result rset{txn.exec(action_sql)};
-      }
-    }
+    ApplyHistory(txn, true /*use_hypopg*/);
```
In justification of my above comment on ApplyHistory, I think this breaks the non-hypopg code path...

```suggestion
    ApplyHistory(txn, use_hypopg);
```

is what you wanted, but I think

```suggestion
    ApplyHistory(txn);
```

is what it should be refactored into.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -483,28 +552,19 @@ std::vector<double> DatabaseState::Returns() const {
     //  Right now, the workload only appears at the very end after all actions are applied.
     const std::vector<ForecastedQuery> &workload = forecast->GetForecastWorkload();
+
+    // This vector is used to record predictions. The vector (if valid) corresponds 1:1
```
Nit: you are declaring this vector and making people read this, even for code paths that have nothing to do with predictions.

I am OK with this for now, but I think we are approaching the point where we will want to write an `EvaluateWorkloadCost` function that calls `EvaluateWorkloadExplainCost`, `EvaluateWorkloadExplainAnalyzeCost`, etc.

Can you add a comment like

```suggestion
    // TODO(WAN): If you find yourself needing to declare more flag-specific variables here,
    // refactor this into an `EvaluateWorkloadCost` function that calls separate costing functions.
    //
    // This vector is used to record predictions. The vector (if valid) corresponds 1:1
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
+}
+
+template<class T>
+std::pair<bool, double> DatabaseState::GetExplainAnalyzeCostUs(T &txn, const std::string &query) const {
```
Out of curiosity, why microseconds when EXPLAIN ANALYZE operates mostly in milliseconds? Can you add the reasoning in a comment as well?

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+      }
+    }
+
+    if (use_hypopg && use_microservice && record_predictions) {
```
It seems odd to be conditioned on `use_hypopg` here; I guess the argument is that you cannot use the microservice without using hypopg.. which is false, just that we don't support that code path yet. I opened a suggestion to have some asserts in the ctor.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -556,10 +653,14 @@ DatabaseGame::DatabaseGame(const GameParameters &params) : Game(kGameType, param
   use_hypopg_ = params.at("use_hypopg").bool_value();
   use_microservice_ = params.at("use_microservice").bool_value();
+  record_predictions_ = params.at("record_predictions").bool_value();
 
```
```suggestion

  // Microservice currently assumes HypoPG. There is no reason this has to be the case.
  SPIEL_CHECK_TRUE(!use_microservice_ || (use_microservice_ && use_hypopg_));
  // Recording currently requires microservice. This is because data is recorded to the microservice.
  SPIEL_CHECK_TRUE(!record_predictions_ || (record_predictions_ && use_microservice_));

```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+    }
+
+    if (use_hypopg && use_microservice && record_predictions) {
+      // Remove all hypothetical indexes
```
```suggestion
      // Remove all hypothetical indexes.
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+      // Remove all hypothetical indexes
+      txn.exec("select hypopg_reset();");
+
+      // Start a sub-transaction. When this sub-transaction goes away, all the actions (which
```
```suggestion
      // Start a sub-transaction. When this sub-transaction goes away, all of the actions that we apply (which
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+      // Start a sub-transaction. When this sub-transaction goes away, all the actions (which
+      // we assume are transactional) will be automatically rolled back.
+      pqxx::subtransaction subtxn(txn);
+      ApplyHistory(subtxn, false);
```
This is where you would want something like

```suggestion
      CustomApplyHistory(subtxn, false);
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+      // we assume are transactional) will be automatically rolled back.
+      pqxx::subtransaction subtxn(txn);
+      ApplyHistory(subtxn, false);
+      std::string action_state = GetAppliedStateRepresentation();
```
See above comments about duplicate functionality, I don't think you need this.

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+      ApplyHistory(subtxn, false);
+      std::string action_state = GetAppliedStateRepresentation();
+
+      for (size_t idx = 0; idx < workload.size(); idx++) {
```
To remind people why you're doing index-based iteration,

```suggestion

      SPIEL_CHECK_EQ(workload.size(), prediction_workloads.size());
      for (size_t idx = 0; idx < workload.size(); idx++) {
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+      std::string action_state = GetAppliedStateRepresentation();
+
+      for (size_t idx = 0; idx < workload.size(); idx++) {
+        // Get the true cost of executing the query with the configuration
```
```suggestion
        // Get the true cost of executing this query with the current configuration.
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+        // Get the true cost of executing the query with the configuration
+        std::pair<bool, double> true_cost = GetExplainAnalyzeCostUs(subtxn, workload[idx].sql_);
+        if (!true_cost.first) {
+          // In this case, the subtransaction has failed (i.e., we hit an integrity key
```
```suggestion
          // In this case, the subtransaction has failed (e.g., we hit an integrity key
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +577,50 @@ std::vector<double> DatabaseState::Returns() const {
+          // violation). For this case, we need to restart the subtransaction.
+          subtxn.~subtransaction();
+          new (&subtxn) pqxx::subtransaction(txn);
+          ApplyHistory(subtxn, false);
```
This has the potential to redo a nontrivial amount of work. A subtransaction is basically a SAVEPOINT, and afaik you can have multiple SAVEPOINTs in a transaction.

lmwnshn replied:
If you start the subtxn _after_ applying the history, and the subtxn is only for executing the current workload query, this all goes away I think?

17zhangw replied:
Yes. This should eliminate unnecessary re-applying of history.

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
 
 
+def connect():
+    # pylint: disable=assigning-non-slot
```
I would just remove this and add the pylint exception, since this is "the flask way" of doing things and I don't expect people to know what a Python slot is anyway.

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
 
+def connect():
+    # pylint: disable=assigning-non-slot
+    db = getattr(g, "__database", None)
```
If you want to be super nitpicky, I think the convention used in Flask is single underscore `_database` (like a "please do not use, internals only" Python member). Unless you wanted the anti name mangling?

```suggestion
    db = getattr(g, "_database", None)
```

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+    # pylint: disable=assigning-non-slot
+    db = getattr(g, "__database", None)
+    if db is None:
+        db = g.__database = sqlite3.connect("inference.db")
```
```suggestion
        db = g._database = sqlite3.connect("inference.db")
```

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+
+@app.teardown_appcontext
+def close_connection(_):
+    db = getattr(g, "__database", None)
```
```suggestion
    db = getattr(g, "_database", None)
```

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+        db.close()
+
+
+def init_db():
```
Since this isn't exposed as a route,
```suggestion
def _init_db():
```

?

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+        db.commit()
+
+
+def get_inference_results(query):
```
```suggestion
def _get_inference_results(query):
```

?

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+
+
+def get_inference_results(query):
+    # This by default returns the data sorted by query_id, then ordered
```
Please use numpy-style docstrings. Most of this information would fall under the `Returns` section.

```
        """
        Blablablablabla.

        Parameters
        ----------
        query : str
            blablablabla

        Returns
        -------
        result_set : what'sthetype
            Most of your text blablabla
```

lmwnshn replied:
And query isn't even really the query right? Is there a reason to make it not the whole query, which would be conceptually simpler? To form queries, people already have to know the schema. We're not bothering to escape and this is local anyway, so SQL injection isn't much of a concern. (not that this protects against SQL injection either)

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+
+    def dict_factory(cursor, row):
+        ret = {}
+        for idx, col in enumerate(cursor.description):
```
What is `col[0]`? I suggest adding separate naming lines here
```
blah = col[0]
blabla = row[idx]
```

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+        "SELECT *, ABS(true_cost - predicted_cost) as cost_diff FROM inference_results " + query_clause
+    )
+    result_set = cursor.fetchall()
+    db.commit()
```
Mostly curiosity: is there a reason to commit the select?

17zhangw replied:
It's not really necessary. Mainly did this just to be consistent (open --> close)

lmwnshn commented on behavior/microservice/app.py:
```diff
@@ -41,10 +42,122 @@ def infer(model_type, ou_type):
+
+@app.route("/")
+def index():
+    results = get_inference_results(None)
```
Seems like you want a default argument on query. The default argument could be the default SQL that gets executed.

lmwnshn commented on behavior/microservice/templates/base.html:
```diff
@@ -0,0 +1,19 @@
+<html>
+  <head>
+    <title>{{ title }}</title>
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.1/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-+0n0xVW2eSR5OomGNYDnhzAbDsOXxcvSN1TPprVMTNDbiYZCxYbOOl7+AMvyTG2x" crossorigin="anonymous">
```
I have mixed opinions on cdn vs just downloading the css and including a minified version. Do you have strong opinions against downloading these to `microservice/static`? I'm thinking about the no-internet airplane workflow right now.

17zhangw replied:
Switched to downloading these to `microservice/static`.

lmwnshn reviewed (changes requested):
Looks good overall! Approach is fine, mostly nits.

17zhangw committed:
Refactor and address comments

17zhangw committed:
Switch from CDN -> static files

lmwnshn commented on pyproject.toml:
```diff
@@ -58,6 +58,9 @@ disable = [
     # defining self.members outside of init. A fix would be welcome.
     "attribute-defined-outside-init",
+
+    # Disable assigning non-slot exception
```
```suggestion
    # Disable assigning non-slot exception.
```

lmwnshn commented on action/selection/open_spiel/database.cc:
```diff
@@ -517,13 +573,57 @@ std::vector<double> DatabaseState::Returns() const {
+      }
+    }
+
+    // This is conditioned on use_hypopg only that we only support recording predictions
```
```suggestion
    // This is conditioned on use_hypopg only because we only support recording predictions
```

lmwnshn reviewed (approved):
LGTM! Thanks for addressing all the nits.

lmwnshn committed:
Fix minor grammar nit.

lmwnshn committed:
Fix minor grammar nit.

lmwnshn commented:
I probably _should_ run it again before merging, but if it breaks we'll find out soon. Not that I expect it to break.

lmwnshn committed:
Change representation of DatabaseState::ToString().

Since this is a single player game, we don't need player information.
Additionally, replace action ID with action SQL for ease of use.
